transform_aes.c

```diff
@@ -219,28 +220,37 @@ static int transop_encode_aes( n2n_trans_op_t * arg,
             encode_uint32( outbuf, &idx, sa->sa_id );
 
-            /* Encrypt the assembly contents and write the ciphertext after the SA. */
-            len = in_len + TRANSOP_AES_NONCE_SIZE;
+           /* The following fields (message authentication, IV) are calculated respectively filled after data encrpytion */
 
-            /* The assembly buffer is a source for encrypting data. The nonce is
+            /* Encrypt the assembly contents and write the ciphertext after the IV. */
+            len = in_len + TRANSOP_AES_DATA_LEN;
+
+            /* The assembly buffer is a source for encrypting data. The data length is
              * written in first followed by the packet payload. The whole
              * contents of assembly are encrypted. */
-            pnonce = (uint32_t *)assembly;
-            *pnonce = rand();
-            memcpy( assembly + TRANSOP_AES_NONCE_SIZE, inbuf, in_len );
+            data_len = (uint16_t *)assembly;
+            *data_len = in_len;
 
-            /* Need at least one encrypted byte at the end for the padding. */
-            len2 = ( (len / AES_BLOCK_SIZE) + 1) * AES_BLOCK_SIZE; /* Round up to next whole AES adding at least one byte. */
-            assembly[ len2-1 ]=(len2-len);
-            traceEvent( TRACE_DEBUG, "padding = %u", assembly[ len2-1 ] );
+	    memcpy( assembly + TRANSOP_AES_DATA_LEN , inbuf, in_len );
+            len2 = ( ( ( len - 1) / AES_BLOCK_SIZE) + 1 ) * AES_BLOCK_SIZE; /* Round up to next whole AES block size */
+
+            RAND_bytes(&iv, 16);  /* get a cryptographically safely randomized initialization vector */
+            /* REVISIT: return value of RAND_bytes function should be 1 to continue cryptographically safe, otherwise exit */
+            memcpy((sa->enc_ivec), iv, 16 ); /* sa->enc_ivec gets changed during CBC encryption procedure, so keep a copy in iv */
 
-            memset( &(sa->enc_ivec), 0, sizeof(N2N_AES_IVEC_SIZE) );
             AES_cbc_encrypt( assembly, /* source */
-                             outbuf + TRANSOP_AES_VER_SIZE + TRANSOP_AES_SA_SIZE, /* dest */
+                             outbuf + TRANSOP_AES_VER_SIZE + TRANSOP_AES_SA_SIZE + N2N_AES_MSGAUTH_SIZE + N2N_AES_IVEC_SIZE, /* dest */
                              len2, /* enc size */
-                             &(sa->enc_key), sa->enc_ivec, 1 /* encrypt */ );
+                             &(sa->enc_key), &(sa->enc_ivec), 1 /* encrypt */ );
+
+            AES_encrypt (iv, /* encrpyt the original initialization vector */
+                         outbuf + TRANSOP_AES_VER_SIZE + TRANSOP_AES_SA_SIZE + N2N_AES_MSGAUTH_SIZE, /* to the buffer */
+                         &(sa->enc_key_2) ); /* using the secondary key */
+            AES_encrypt (outbuf + len2 - AES_BLOCK_SIZE, /* encrypt the last cipher block again */
+                         outbuf + TRANSOP_AES_VER_SIZE + TRANSOP_AES_SA_SIZE, /* to use it as message authentication code */
```

Comment:
Please provide links to online resources/explain why performing the operation on the last block is enough for the authentication of the whole packet

Reply:
As CBC always uses the cipher of the block before to encrypt the next one, so every cipher always depends on all block before. Therefore, the last block is a fingerprint of the whole encrypted data. Anyway, we need to encrypt it again using another key to prevent an attack described [here](https://en.wikipedia.org/wiki/CBC-MAC#Using_the_same_key_for_encryption_and_authentication).